src/Analyser/NodeScopeResolver.php

```diff
@@ -1913,7 +1913,14 @@ static function (?Type $offsetType, Type $valueType, bool $optional) use (&$arra
 								return;
 							}
 
-							$arrayType = TypeCombinator::union(...TypeUtils::getArrays($arrayType));
+							// Remove NonEmptyArrayType
+							$arrayType = TypeCombinator::intersect(
+								new ArrayType($arrayType->getIterableKeyType(), $arrayType->getIterableValueType()),
+								...array_filter(
```

Comment:
I don't like this isn't a list. Please wrap it in `array_values` too.

Reply:
And is there actually an accessorry type we'd like to preserve here?

Reply:
most likely not. it just felt wrong too remove potentially too much. but this "potentially" is something that doesn't even exist yet. e.g. even having a hasOffset accessory implies that the original array is non-empty and then this code isn't even triggered.
I guess it makes more sense to remove the complexity here again? :)

Reply:
Yes please :) With the unknown future type, we don't know if we'd want to preserve it here or not...

Reply:
rather not optimise for unknown future types :D
gone again, thx for the patience. I'll call it a day. looking forward to some HasOffsetValue niceness btw